### PR TITLE
Long filename file create/write

### PR DESCRIPTION
**Pull request by GHGiampy**
### Description

Marlin already supports reading long filenames but cannot write them. This PR enables support for writing long file names (LFN).
It can be useful in combination with the PR #23462 to upload firmware on cards that require a long file name to flash the printer.
It is not a replacement of the FAT driver but a custom implementation of the missing parts to handle the LFN support, so it needs testing even if it works fine on my printer.
Supports:
- Write - Tested (Read was already supported)
- Delete - Tested
- Rename - Not tested but internally uses the "Write" function which works
- MkDir - Not tested but internally uses "Write" function which works
- GCodes: M20, M23, M28, M29, M30, M33 - Tested

This PR also update the upload script to use long or short filenames reflecting the `LONG_FILENAME_HOST_SUPPORT` define and add a custom flag for use an 8.3 random filename (if needed)

### Requirements

Board with USB connection

### Benefits

Allows the use of long filenames.
Allows boards with long firmware filename to be updated via binary_protocol/USB.

### Configurations

N/A

### Related Issues

N/A

### CODE REVIEW

File: buildroot/share/scripts/upload.py [hunk 1]
```diff
@@ -124,6 +129,7 @@ def _RemoveFirmwareFile(FirmwareFile):
     marlin_board_info_name = _GetMarlinEnv(MarlinEnv, 'BOARD_INFO_NAME')
     marlin_board_custom_build_flags = _GetMarlinEnv(MarlinEnv, 'BOARD_CUSTOM_BUILD_FLAGS')
     marlin_firmware_bin = _GetMarlinEnv(MarlinEnv, 'FIRMWARE_BIN')
+    marlin_long_filename_host_support = _GetMarlinEnv(MarlinEnv, 'LONG_FILENAME_HOST_SUPPORT') is not None
```
Comment (thinkyhead) [inline buildroot/share/scripts/upload.py:132]: This configuration option specifically signifies the presence of `M33`.

Comment (GHGiampy) [reply buildroot/share/scripts/upload.py:132]: It is also used in M20 to retrieve the LFN list with 'L' flag

Comment (thinkyhead) [reply buildroot/share/scripts/upload.py:132]: Hmm, so it does.

**Comment by thinkyhead** [comment]
The new additions supporting the creation of long filenames need to be optional so that the base build size for SD is not increased for all users when SD is enabled. I'm adding a new option to this and starting on the work to make the new code only compile when this option is enabled. It will take some time, but I should have changes to push pretty soon.
```cpp
  // This allows Marlin to create files with long filenames
  //#define LONG_FILENAME_WRITE_SUPPORT
```

**Comment by thinkyhead** [comment]
Using `mftest -tmega -n1 -y` as the base, the firmware build size with the new option is ~2100 bytes larger than without.

**Comment by GHGiampy** [comment]
It's a lot of program space for board with low memory! I did't realize it.
I tested with all combinations of switches, it works well.

**Comment by thinkyhead** [comment]
How well does it work for long filenames with spaces (` `), quotes (`"`), backslashes (`\`), and single quotes (`'`) in them? Do we need to enable the `GCODE_QUOTED_STRINGS` option for these things to work properly? Should the G-code parser also be extended to treat the backslash as a special character, even when `GCODE_QUOTED_STRINGS` is not enabled?

File: Marlin/src/gcode/host/M115.cpp [hunk 2]
```diff
@@ -154,6 +154,12 @@ void GcodeSuite::M115() {
     // LONG_FILENAME_HOST_SUPPORT (M33)
     cap_line(F("LONG_FILENAME"), ENABLED(LONG_FILENAME_HOST_SUPPORT));
 
+    // LONG_FILENAME_WRITE_SUPPORT (M20, M23, M28, M29, M30, M33, ...)
+    cap_line(F("LFN_WRITE"), ENABLED(LONG_FILENAME_WRITE_SUPPORT));
+
+    // CUSTOM_FIRMWARE_UPLOAD (M20 F)
+    cap_line(F("CUSTOM_FIRMWARE_UPLOAD"), ENABLED(CUSTOM_FIRMWARE_UPLOAD));
+
```
Comment (thinkyhead) [inline Marlin/src/gcode/host/M115.cpp:162]: …added to https://reprap.org/wiki/Firmware_Capabilities_Protocol

**Comment by GHGiampy** [comment]
> How well does it work for long filenames with spaces (` `), quotes (`"`), backslashes (`\`), and single quotes (`'`) in them? Do we need to enable the `GCODE_QUOTED_STRINGS` option for these things to work properly? Should the G-code parser also be extended to treat the backslash as a special character, even when `GCODE_QUOTED_STRINGS` is not enabled?

Currently the parser of `M28, M30, M32, and M33` takes the whole string, so spaces (` `) and single quotes (`'`) works out of the box, altough a parser should stop at 1st space and get the token, if not quoted and `GCODE_QUOTED_STRINGS` disabled.
Other reserved characters (`0x00-0x1F,0x7F,",*,/,:,<,>,?,\,|`) result in an open fail as per https://en.wikipedia.org/wiki/Filename#Comparison_of_filename_limitations at VFat section.
The only exception is in the parser of `M23` for the Simplify3D which stops at the 1st space to strip uneeded 'size' info.
Paradoxically only the parser of `M23` is working as it should with `GCODE_QUOTED_STRINGS` disabled (conversely it doesn't work if a quoted string is passed and `GCODE_QUOTED_STRINGS` is enabled), but the custom code is in `M23.cpp` and not in the `parser.cpp`.
I think that a good rework of the parser and the GCODE parameters is needed but it could break other functionalities at this stage.

**Pull request merged**